File: app/src/main/java/com/sd/demo/visible_level/MainActivity.java
package com.sd.demo.visible_level;

import android.os.Bundle;
import android.util.Log;
import android.widget.RadioGroup;

import androidx.appcompat.app.AppCompatActivity;

import com.sd.demo.visible_level.appview.HomeView;
import com.sd.demo.visible_level.appview.LiveView;
import com.sd.demo.visible_level.appview.MeView;
import com.sd.demo.visible_level.databinding.ActivityMainBinding;
import com.sd.demo.visible_level.level_home.HomeLevel;
import com.sd.lib.vlevel.FVisibleLevel;

public class MainActivity extends AppCompatActivity
{
    public static final String TAG = MainActivity.class.getSimpleName();

    private static final String LEVEL_ITEM_HOME = HomeView.class.getName();
    private static final String LEVEL_ITEM_LIVE = LiveView.class.getName();
    private static final String LEVEL_ITEM_ME = MeView.class.getName();

    private ActivityMainBinding mBinding;

    private HomeView mHomeView;
    private LiveView mLiveView;
    private MeView mMeView;

    private final FVisibleLevel mVisibleLevel = FVisibleLevel.get(HomeLevel.class);

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        mBinding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());

        mHomeView = new HomeView(this);
        mLiveView = new LiveView(this);
        mMeView = new MeView(this);

        mVisibleLevel.addVisibilityCallback(mVisibilityCallback);
        mVisibleLevel.getItem(HomeLevel.ItemHome.class).addVisibilityCallback(mHomeView);
        mVisibleLevel.getItem(HomeLevel.ItemLive.class).addVisibilityCallback(mLiveView);
        mVisibleLevel.getItem(HomeLevel.ItemMe.class).addVisibilityCallback(mMeView);

        mBinding.radioMenu.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener()
        {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId)
            {
                mBinding.flContainer.removeAllViews();
                switch (checkedId)
                {
                    case R.id.btn_home:
                        mVisibleLevel.visibleItem(HomeLevel.ItemHome.class);
                        mBinding.flContainer.addView(mHomeView);
                        break;
                    case R.id.btn_live:
                        mVisibleLevel.visibleItem(HomeLevel.ItemLive.class);
                        mBinding.flContainer.addView(mLiveView);
                        break;
                    case R.id.btn_me:
                        mVisibleLevel.visibleItem(HomeLevel.ItemMe.class);
                        mBinding.flContainer.addView(mMeView);
                        break;
                    default:
                        break;
                }
            }
        });

        mBinding.radioMenu.check(R.id.btn_home);
    }

    private final FVisibleLevel.VisibilityCallback mVisibilityCallback = new FVisibleLevel.VisibilityCallback()
    {
        @Override
        public void onLevelVisibilityChanged(boolean visible, FVisibleLevel level)
        {
            Log.i(TAG, "onLevelVisibilityChanged visible:" + visible + " level:" + level);
        }
    };
}

File: app/src/main/java/com/sd/demo/visible_level/MainActivity.kt
package com.sd.demo.visible_level

import android.os.Bundle
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.sd.demo.visible_level.databinding.ActivityMainBinding
import com.sd.demo.visible_level.level.LevelHome
import com.sd.lib.vlevel.FVisibleLevel
import com.sd.lib.vlevel.fVisibleLevel
import com.sd.lib.vlevel.fVisibleLevelRemove

class MainActivity : AppCompatActivity() {

    init {
        FVisibleLevel.isDebug = true
    }

    private val _binding by lazy { ActivityMainBinding.inflate(layoutInflater) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(_binding.root)

        _binding.radioMenu.setOnCheckedChangeListener { _, checkedId ->
            when (checkedId) {
                R.id.btn_home -> {
                    fVisibleLevel<LevelHome>().setCurrentItem(LevelHome.Home)
                    selectView(R.id.view_home)
                }
                R.id.btn_live -> {
                    fVisibleLevel<LevelHome>().setCurrentItem(LevelHome.Live)
                    selectView(R.id.view_live)
                }
                R.id.btn_me -> {
                    fVisibleLevel<LevelHome>().setCurrentItem(LevelHome.Me)
                    selectView(R.id.view_me)
                }
                else -> {}
            }
        }
        _binding.radioMenu.check(R.id.btn_home)
    }

    private fun selectView(id: Int) {
        repeat(_binding.flContainer.childCount) { index ->
            val child = _binding.flContainer.getChildAt(index)
            if (child?.id == id) {
                child.visibility = View.VISIBLE
            } else {
                child.visibility = View.GONE
            }
        }
    }

    override fun onStart() {
        super.onStart()
        fVisibleLevel<LevelHome>().setVisible(true)
    }

    override fun onStop() {
        super.onStop()
        fVisibleLevel<LevelHome>().setVisible(false)
    }

    override fun onDestroy() {
        super.onDestroy()
        fVisibleLevelRemove<LevelHome>()
    }
}

inline fun logMsg(block: () -> String) {
    Log.i("visible-level-demo", block())
}